[#9560] Improve L&P test process

amrut-prabhu opened the pull request:
Part of #9560

**Outline of solution**:

- Instead of specifying `JSON_DATA_PATH` and `CSV_CONFIG_PATH`, now the tests use the class name. For example, `StudentProfileLNPTest.json` and `StudentProfileLNPTestConfig.csv`
- The created data is stored in the order of creation (numerical order) in the json and csv files.
- The `lnpTests` gradle task now runs every time it's executed, even if no code changes have been made.
- The gradle build output is now able to show console log output from the test. Currently, it is limited to 2 lines per test (1 line for summary, and the other one _can_ be for the link to a HTML report)
    ![image](https://user-images.githubusercontent.com/29021586/56094471-5eb76d80-5f07-11e9-892f-b923bbb8bf52.png)
- Refactored the `JMeterConfig` class (and renamed it to `JMeterElements`), and removed the createTestPlan() method entirely (as well as most of the related getter methods). Test plan structures are now created within the individual test cases. This provides more flexibility and looks a bit cleaner overall (in my opinion)

amrut-prabhu commented:
@RonakLakhotia @jacoblipech Ready for review!

RonakLakhotia commented on build.gradle:
```diff
@@ -589,7 +590,21 @@ task lnpTests(type: Test) {
+    beforeTest { TestDescriptor td -> outputCache.clear() } // clear output cache before the test starts
+    onOutput { TestDescriptor td, TestOutputEvent toe -> // put output in the cache
+        outputCache.add(toe.getMessage())
+        while (outputCache.size() > 2) outputCache.remove() // keep only last 2 result lines; if more than 3 lines -> drop first
```
Not entirely sure if I follow the reasoning here? Could you help me understand this better

amrut-prabhu replied:
This is just a precaution to prevent the build output from becoming messy. This will show the last 2 lines that are logged to the console, in the `gradlew lnpTests` build output. 
Ideally, we only want the JMeter test results to get displayed here. 

I haven't been able to extensively test this and there is an assumption made here, but this is the best workaround that I could find. The assumption is that nothing else is printed after we print the JMeter results summary.

amrut-prabhu replied:
> if more than 3 lines

this is a typo. Should be `2` 😬

amrut-prabhu commented:
Ready for review.
Refactored the `JmeterConfig` class and removed the `createTestPlan()` method entirely (as well as most of the related getter methods).

RonakLakhotia commented on src/e2e/java/teammates/e2e/cases/lnp/BaseLNPTestCase.java:
```diff
@@ -45,62 +44,39 @@
 
-    private String getPathToTestDataFile(String fileName) {
-        return TestProperties.LNP_TEST_DATA_FOLDER + fileName;
+    protected String getPathToTestDataFile(String fileName) {
```
Missing java docs?

amrut-prabhu replied:
Updated!

RonakLakhotia commented on src/e2e/java/teammates/e2e/cases/lnp/BaseLNPTestCase.java:
```diff
@@ -45,62 +44,39 @@
+        return "/" + getClass().getSimpleName() + ".jtl";
     }
 
     @Override
```
Same here

amrut-prabhu replied:
It's present above

RonakLakhotia commented on src/e2e/java/teammates/e2e/cases/lnp/StudentProfileLNPTest.java:
```diff
@@ -108,29 +104,9 @@ protected LNPTestData getTestData() {
             @Override
             protected Map<String, StudentProfileAttributes> generateProfiles() {
-                Map<String, StudentProfileAttributes> profiles = new HashMap<>();
+                Map<String, StudentProfileAttributes> profiles = new LinkedHashMap<>();
```
Sorry, can you clarify why you changed this from `HashMap` to `LinkedHashMap`

amrut-prabhu replied:
The created data is stored in the order of creation (numerical order) in the json and csv files. Pretty minor, but is more intuitive to us humans when we are debugging with the data

RonakLakhotia commented on src/e2e/java/teammates/e2e/util/JMeterElements.java:
```diff
@@ -0,0 +1,269 @@
+    }
+
+    /**
+     * Returns a HTTP Request element that configured to login to a TEAMMATES instance.
```
* that is configured

RonakLakhotia commented on src/e2e/java/teammates/e2e/util/JMeterElements.java:
```diff
@@ -0,0 +1,269 @@
+    }
+
+    /**
+     * Returns a Once Only Controller element that process the controller(s) inside it only once per thread.
```
*that processes

RonakLakhotia commented on src/e2e/java/teammates/e2e/util/JMeterElements.java:
```diff
@@ -0,0 +1,269 @@
+    /**
+     * Overloaded method that returns a HTTP Request element with a GET method to the endpoint specified by {@path}.
+     */
+    public static HTTPSamplerProxy httpSampler(String path) {
```
change the method name here?

RonakLakhotia commented on src/e2e/java/teammates/e2e/util/JMeterElements.java:
```diff
@@ -0,0 +1,269 @@
+    /**
+     * Returns a HTTP Cookie Manager element.
+     */
+    public static CookieManager cookieManager() {
```
Since the method seems to be a getter, you might want to rename it as `getCookieManager` similar to `getJsonData`

amrut-prabhu replied:
Yeah, I thought about that but decided against it since all the methods in this utility class return TestElements. I felt it would be better to just name the methods after the elements

Also, this isn't technically a getter, but I understand what you mean :stuck_out_tongue:

amrut-prabhu commented:
@wkurniawan07 Requesting your review
